piquasso/api/simulator.py

```diff
@@ -260,11 +322,18 @@ def execute_instructions(
         self._validate_instructions(instructions)
 
         if initial_state is not None:
-            self._validate_state(initial_state)
+            self._validate_initial_state(initial_state)
             state = initial_state.copy()
         else:
             state = self.create_initial_state()
 
+        is_many_measurement = self._detect_many_measurement_scenario(instructions)
+
+        if is_many_measurement:
+            return self._execute_many_measurement_instructions(
+                state, instructions, shots
+            )
+
         result = Result(state=state)
 
         for instruction in instructions:
```

Comment:
We discussed this briefly, imho a good design could be:
1. Outer function that takes list of instructions and calls one of the following inner functions:
2. a) Function executes instructions with one measurement;
2. b) Function executes instructions with multiple measurements.

Then there is a function that simply takes a list of instructions and executes those and gives back the samples (not the Result object yet). 2a) and 2b) should use this function and e.g., the definition of 2a) may be quite trivial. Also, Result objects should be created only once inside of function 1. very likely, following the idea that: "once we have all the results, we put them into a Result object".

Reply:
Indeed, it would be a better to use your suggested pattern, I would definitely avoid duplication here.